### PR TITLE
Allow resizing Strings from GDExtension

### PR DESCRIPTION
**Pull request by dsnopek**
This adds a new function to `gdextension_interface.h` to allow us to implement `String::resize()` in godot-cpp per issue https://github.com/godotengine/godot-cpp/issues/1141

This is added to `gdextension_inteface.h` rather than `ClassDB`, because we don't want scripting to have access to it. `String::resize()` only makes sense if you can access `String::ptrw()` and directly modify the data at the pointer, which would be impossible from GDScript.

This originally came up when trying to compile the text_server_adv module in Godot as a GDExtension. See: https://github.com/godotengine/godot/pull/77532

### CODE REVIEW

**Review by bruvzg — COMMENTED** [review]
Is it necessary to expose it directly, adding `bind_string_method(resize, sarray("size"), varray());` to the `variant_call.cpp` to expose existing method should work. The other string functions are in the interface due to pointer arguments, but in this case it's only `int` size.

**Comment by bruvzg** [comment]
> This is added to gdextension_inteface.h rather than ClassDB, because we don't want scripting to have access to it.

Why? It's exposed for the `Array`s, what's the difference, `String` have writable `[]` operator exposed, which is the same as `ptrw()` ?

**Comment by dsnopek** [comment]
> Why? It's exposed for the `Array`s, what's the difference, `String` have writable `[]` operator exposed, which is the same as `ptrw()` ?

Ah, that's a good point!

When I get a chance, I'll make a new PR that exposes it normally...

**Comment by dsnopek** [comment]
I've open PR https://github.com/godotengine/godot/pull/79177 which exposes `String.resize()` the normal way.

Given that `String.resize()` takes the desired array size of the string (ie. including the null terminator), I'm now less sure that we really want this exposed to GDScript, at least without some additional changes to make it more user-friendly. (Some possible options listed over on the other PR.)

**Comment by AThousandShips** [comment]
As discussed in #79177 this seems to be the better way to handle this, assuming it doesn't suffer from the same mutability issues

**Review by bruvzg — APPROVED** [review]

File: core/extension/gdextension_interface.h [hunk 1]
```diff
@@ -1526,6 +1526,19 @@ typedef void (*GDExtensionInterfaceStringOperatorPlusEqWcstr)(GDExtensionStringP
  */
 typedef void (*GDExtensionInterfaceStringOperatorPlusEqC32str)(GDExtensionStringPtr p_self, const char32_t *p_b);
 
+/**
+ * @name string_resize
+ * @since 4.2
+ *
+ * Resizes the underlying string data to the given number of characters.
```
Comment (AThousandShips) [inline core/extension/gdextension_interface.h:1533]: Might want to specify that the string needs to be null terminated manually

Comment (dsnopek) [reply core/extension/gdextension_interface.h:1533]: Ah, yeah, that's a good idea! I've added a note about the null terminating character in my latest push

**Review by AThousandShips — APPROVED** [review]

**Comment by dsnopek** [comment]
Discussed at the GDExtension meeting, and we think it makes sense to add this, however, it would be good to add some additional warnings in the documentation about this being a potentially dangerous operation.

**Comment by YuriSizov** [comment]
> it would be good to add some additional warnings in the documentation about this being a potentially dangerous operation.

Do you mean the online documentation? Or is there something you still want to add to the class references (and therefore need to update this PR)?

**Comment by dsnopek** [comment]
> Do you mean the online documentation?

We were talking about the docs in the comments in `gdextension_interface.h`. I've already done a push since the meeting that adds this:

```c++
 * Warning: This is an error-prone operation - only use it if there's no other
 * efficient way to accomplish your goal.
```

**Comment by YuriSizov** [comment]
Thanks!

**Pull request merged**